cmd/clusterctl/client/tree/discovery.go

```diff
@@ -69,10 +70,14 @@ func Discovery(ctx context.Context, c client.Client, namespace, name string, opt
 	}
 
 	// Adds control plane
-	controlPLane, err := external.Get(ctx, c, cluster.Spec.ControlPlaneRef, cluster.Namespace)
-	if err == nil {
-		tree.Add(cluster, controlPLane, ObjectMetaName("ControlPlane"), GroupingObject(true))
+	if cluster.Spec.ControlPlaneRef == nil {
+		return nil, errors.Errorf("controlplane ref cannot be nil")
```

Comment:
Following up on https://github.com/kubernetes-sigs/cluster-api/pull/4343#discussion_r597872175 and https://github.com/kubernetes-sigs/cluster-api/pull/4343#discussion_r598856834 I opted to fail early with an error message. Note that as it stands `Test_Discovery` is failing because it always assumes:

```
g.Expect(tree).ToNot(BeNil())
g.Expect(err).ToNot(HaveOccurred())
```

I haven't modified anything further to fix this and wanted to open it up for review first.